src/devices/somfy.c

```diff
@@ -0,0 +1,194 @@
+    "? (0)",
+    "My (1)",
+    "Up (2)",
+    "My + Up (3)",
+    "Down (4)",
+    "My + Down (5)",
+    "Up + Down (6)",
+    "? (7)",
+    "Prog (8)",
+    "Sun + Flag (9)",
+    "Flag (10)",
+    "? (11)",
+    "? (12)",
+    "? (13)",
+    "? (14)",
+    "? (15)",
+};
+
+static int somfy_rts_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    data_t *data;
+
+    uint8_t is_retransmission = 0;
+    uint8_t decode_row = 0;
+    uint8_t data_start = 0;
+    const uint8_t *preamble_pattern;
+    unsigned preamble_pattern_bit_length = 0;
+
+    bitbuffer_t bitbuffer_decoded = { 0 };
+
+    uint8_t message_bytes[7] = { 0 };
+    uint8_t chksum_calc = 0;
+    uint8_t chksum = 0;
+    uint16_t counter = 0;
+    char address[7];
+    uint8_t control = 0;
+    uint8_t seed = 0;
+
+    for (int i = 0; i < bitbuffer->num_rows; i++) {
+        if (bitbuffer->bits_per_row[i] > 170) {
+            is_retransmission = 1;
+            decode_row = i;
+            data_start = 65;
+            preamble_pattern = (const uint8_t *) "\xf0\xf0\xf0\xf0\xf0\xf0\xf0\xff";
+            preamble_pattern_bit_length = 64;
+            break;
+        } else if (bitbuffer->bits_per_row[i] > 130) {
+            is_retransmission = 0;
+            decode_row = i;
+            data_start = 25;
+            preamble_pattern = (const uint8_t *) "\xf0\xf0\xff";
+            preamble_pattern_bit_length = 24;
+            break;
+        }
+    }
+
+    if (data_start == 0)
+        return DECODE_FAIL_SANITY;
+
+    if (bitbuffer_search(bitbuffer, decode_row, 0, preamble_pattern, preamble_pattern_bit_length) != 0)
+        return DECODE_FAIL_SANITY;
+
+    if (bitbuffer_manchester_decode(bitbuffer, decode_row, data_start, &bitbuffer_decoded, 56) - data_start < 56)
+        if (bitbuffer_manchester_decode(bitbuffer, decode_row, data_start - 1, &bitbuffer_decoded, 56) - data_start + 1 < 56)
+            return DECODE_FAIL_MIC;
+
+    bitbuffer_extract_bytes(&bitbuffer_decoded, 0, 0, message_bytes, 56);
+
+    // descramble
+    for (int i = 6; i > 0; i--)
+        message_bytes[i] = message_bytes[i] ^ message_bytes[i-1];
+
+    // calculate checksum
+    chksum_calc = xor_bytes(message_bytes, 7);
+    chksum_calc = (chksum_calc & 0xf) ^ (chksum_calc >> 4); // fold to nibble
+
+    // Fail if checksum is incorrect
+    if (chksum_calc != 0)
+        return DECODE_FAIL_MIC;
+
+    // extract seed
+    seed = message_bytes[0];
+
+    // extract control
+    control = (message_bytes[1] & 0xf0) >> 4;
+
+    // extract checksum
+    chksum = message_bytes[1] & 0xf;
+
+    // extract counter independent of system's byte order
+    counter = message_bytes[3] + message_bytes[2] * 255;
+
+    // extract address bytes into hexstring
+    snprintf(address, sizeof(address), "%02X%02X%02X", message_bytes[4], message_bytes[5], message_bytes[6]);
+
+    /* clang-format off */
+    data = data_make(
+            "model",          "",               DATA_STRING, "Somfy-RTS",
```

Comment:
we always try hard to define an id (it's used by default in e.g. mqtt), is address possibly the id here? or is that a variable "target address"?

Reply:
Address is a fixed address of the transmitter. We could use this as an id. Should I just replace address by id?

Reply:
Btw. do you have some documentation where you describe all of those requirements? Such as always defining an id, ...

Reply:
That would be great, but it needs to be plain `int`.
We are a bit light on documentation, this explains some expectations: https://triq.org/rtl_433/DATA_FORMAT.html

Reply:
I can convert the address into an integer, no problem.

Reply:
but according to the documentation it should be a string.

Reply:
We generally don't offer a specific "presentation" of data, just raw values. There is sometimes a valid need to do it different, but with 200 decoders we need to consider carefully to preserve some sanity.

Reply:
How large can the id be? the address is a 24 bit value. However, after using it as Id, in the output, I only see an 8 bit number.

Reply:
31 bit should always be safe, so `(message_bytes[4] << 16) | (message_bytes[5] << 8) | message_bytes[6]` should work.

Reply:
you are right, I should not add 8 bit number and expect a 32 bit output ;-). Now it works.